src/lj_str.c

```diff
@@ -161,34 +199,101 @@ GCstr *lj_str_new(lua_State *L, const char *str, size_t lenx)
   h ^= b; h -= lj_rol(b, 16);
   /* Check if the string has already been interned. */
   o = gcref(g->strhash[h & g->strmask]);
+#if LUAJIT_SMART_STRINGS
+/* Regular closed-addressing hash table with fill factor 100% has
+ * "average maximum" collision chain near 6 (probability of longer
+ * collision chain is low). Collision chain of length 40 is certain
+ * sign of bad behaviour, so lets react on it.*/
+#define max_collisions 40
+#define inc_collision_soft() (collisions++)
+/* If same hash is present, grow count faster */
+#define inc_collision_hard() (collisions+=1+(len>>4), 1)
+#else
+#define inc_collision_hard() (1)
+#define inc_collision_soft()
+#endif
   if (LJ_LIKELY((((uintptr_t)str+len-1) & (LJ_PAGESIZE-1)) <= LJ_PAGESIZE-4)) {
     while (o != NULL) {
       GCstr *sx = gco2str(o);
-      if (sx->len == len && str_fastcmp(str, strdata(sx), len) == 0) {
+      if (sx->hash == h && sx->len == len && inc_collision_hard() &&
+                      str_fastcmp(str, strdata(sx), len) == 0) {
 	/* Resurrect if dead. Can only happen with fixstring() (keywords). */
 	if (isdead(g, o)) flipwhite(o);
 	return sx;  /* Return existing string. */
       }
       o = gcnext(o);
+      inc_collision_soft();
     }
   } else {  /* Slow path: end of string is too close to a page boundary. */
     while (o != NULL) {
       GCstr *sx = gco2str(o);
-      if (sx->len == len && memcmp(str, strdata(sx), len) == 0) {
+      if (sx->hash == h && sx->len == len && inc_collision_hard() &&
+                      memcmp(str, strdata(sx), len) == 0) {
 	/* Resurrect if dead. Can only happen with fixstring() (keywords). */
 	if (isdead(g, o)) flipwhite(o);
 	return sx;  /* Return existing string. */
       }
       o = gcnext(o);
+      inc_collision_soft();
+    }
+  }
+#if LUAJIT_SMART_STRINGS
+  /* "fast" hash function fully consumes all bytes of strings <= 12 bytes */
+  if (len > 12) {
+    /* High 12 bits of "fast hashsum" is used for probalistic filter.
+     * High 6 bits are copied to from "fast" to "full hashsum",
+     * and 19-25 bits are stored into s->strflags. */
```

Comment:
The bloom filter is keyed with the high 12 bits of the "fast" hash sum. The filter is rebuilt during GC cycle. It's beneficial to have these bits readily available and avoid hash sum recalculation during GC. High 6 bits are included in the "full" hash sum, and bits 19-25 are stored in s->strflags.